plugins/scaffolder-backend/src/scaffolder/stages/publish/bitbucket.ts

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright 2020 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { PublisherBase, PublisherOptions, PublisherResult } from './types';
+import { pushToRemoteUserPass } from './helpers';
+import { RequiredTemplateValues } from '../templater';
+import { Config, JsonValue } from '../../../../../../packages/config/src';
+import fetch from 'cross-fetch';
+import {
+  BitbucketIntegrationConfig,
+  readBitbucketIntegrationConfigs,
+} from '../../../../../../packages/integration/src';
+
+export class BitbucketPublisher implements PublisherBase {
+  private readonly integrations: BitbucketIntegrationConfig[];
+  private readonly token: string;
+  private readonly host: string;
+
+  constructor(config: Config) {
+    this.token = '';
+    this.host = '';
+    this.integrations = readBitbucketIntegrationConfigs(
+      config.getOptionalConfigArray('integrations.bitbucket') ?? [],
+    );
+  }
+
+  async publish({
+    values,
+    directory,
+  }: PublisherOptions): Promise<PublisherResult> {
+    const result = await this.createRemote(values);
+
+    await pushToRemoteUserPass(
+      directory,
+      result.remoteUrl,
+      this.getUsername(''),
+      this.getPassword('') || '',
+    );
+    return result;
+  }
+
+  private async createRemote(
```

Comment:
@benjdlambert To clarify; this is where I need to be able to know the host so I can call the correct Bitbucket endpoint with the credentials specified for that given host.

Reply:
Exactly, so I've taken some inspiration from how the gitlab preparer used to do it to form something of a patter for all the preparers and publishers. You can take a little look here, I just pushed my work before the holidays to a draft PR here #3900

Reply:
Basically parse the git-url and grab the host, and then call an function to grab a token for the host which will be replaced entirely with the integration config eventually but right now wanting to support both.

Reply:
I see now however, that the publishers have a little bit of a strange abstraction as they get given a token and a client right and some other config, which doesn't really work with the integrations pattern unless the processor would build the client necessary for the right host. I'm thinking that we should move the `publishers` to be created with the config like the preparers, and it can create its own client for github with the correct hostname and apiBaseUrl rather than it being passed in. It's not great in terms of SOLID as it's dependencies are created inside. but I'd rather the job processor be simple as that's the extension point in future releases for the scaffolder. Thoughts @freben @Rugvip

Reply:
I'd rather the publishers become much more like this: https://github.com/backstage/backstage/pull/3900/files#diff-803d91fb044e29fbfdfe04a476d22ff842019eeb0b92a3d05b3da5246e431388R36-R55

Reply:
And then have something similar like this for all the config options to the client needed to create the remote repo https://github.com/backstage/backstage/pull/3900/files#diff-803d91fb044e29fbfdfe04a476d22ff842019eeb0b92a3d05b3da5246e431388R101-R106

Reply:
And when I (or somebody else) gets around to it, we should instead all get an `integrations: Integrations` injected instead, so you can do `integrations.github.byUrl(url).config.token` or whatever.

Reply:
@freben exactly, that's a step in the right direction. Right now it's in a little bit of a weird state, as it's starting to feel like the wrong abstraction in the preparers entirely. They shouldn't get given a `token` and a `client` and `config`.

Reply:
@benjdlambert @freben thanks for the comments.
Would you suggest that we implement the bitbucket scaffolder feature in stages? so the fist stage is to use the `scaffolder.bitbucket` key for configuring the scaffolder as the other integrations currently do it and then refactor when the discussed changes are implemented? or do you prefer another approach?

Reply:
Yes, don't block on us.

Reply:
Agreed, @benjdlambert ?

Reply:
Yep that makes a lot of sense @goober. That would save a lot of time I think. 👍